rp2: Do not format the file system in response to Ctrl+C.

DavidEGrayson opened the pull request:
I noticed that the @mu-editor sends Ctrl+A followed by Ctrl+D to reboot the board, and then it very quickly sends a Ctrl+C character, interrupting the code in `_boot.py` or `_boot_fat.py` while it is trying to mount the root file system.  The `except:` clause in those files catches the `KeyboardInterrupt` and destroys the existing file system.

~~This pull request fixes the problem by only catching exceptions that inherit from the `Exception` class, which does not include `KeyboardInterrupt`.  In the future, maybe the `except` clause should be restricted further, since formatting the file system is so destructive and it only makes sense to do it if we have confirmed that the file system is invalid.~~

~~Note that this pull request is not sufficient to actually get Mu to work with an RP2040: the errant Ctrl+C will interrupt the booting process and leave the interpreter in a state where it does not have a root file system.~~

dpgeorge commented:
It might be worth trying to fix Mu instead, because `mpremote` (and `pyboard.py`) both handle this situation without any problem.  Ideally Mu should follow the logic in `pyboard.py` to interact with a board.

Can you link to the code in Mu that does the ctrl-C?

DavidEGrayson commented:
I have not started looking at the Mu code, but here is the beginning of the USB traffic generated when I press the "Run" button in Mu, as captured with a Beagle USB 12.  The Ctrl+C that causes the problem is on line 483:

![image](https://user-images.githubusercontent.com/466764/221073863-ecd3ad93-7c35-4bd0-8bc4-74157f1ac83e.png)

Even if Mu were fixed tomorrow, I think this pull request still has value because users will encounter old versions of Mu and they could also encounter other software with similar bugs, and they will probably not want to not lose their filesystem when that happens.

pdg137 commented:
> Can you link to the code in Mu that does the ctrl-C?

I think some of the offending code is here:

https://github.com/mu-editor/mu/blob/master/mu/modes/base.py#L172-L189

DavidEGrayson commented:
I just force-pushed a few times, undoing my changes to `_boot.py` and `_boot_fat.py` and implementing a better solution instead.  This pull request ignores Ctrl+C while we are executing the frozen boot code in `_boot.py` or `_boot_fat.py`, preventing the `KeyboardInterrupt` from being raised.  This makes the RP2 port closer to a normal Linux computer: you can no longer press Ctrl+C during the boot process and accidentally get into a state where the root file system isn't mounted.  It also means that the Mu editor will work fine with RP2 devices without needing to be patched.

robert-hh commented:
Wouldn't it be easier to disable Ctrl-C in _boot.py? It just needs to be disabled. Re-enabling happens at the next call to run a script or REPL prompt..

DavidEGrayson commented:
I believe there could still be a window where the boot process can be interrupted, even if the first line of the script is `import micropython` and the second line is `micropython.kbd_intr(-1)`, but I don't know that much about the scheduler works.

robert-hh commented:
Sure. But the question is still
a) should we compensate for the misbehavior of a single PC application, like MU Editor, or
b) should we care about a possible data loss during boot, if Ctrl-C happens at the wrong time.
and what is the simplest fix.
Misbehavior of e.g. the mu-Editor should be fixed there. And as you said, the reformatted file system of the board happens because the `except` includes KeybordInterrupt. That could be avoided by either disabling interrupts in _boot.py or making the except case more specific to not having a file system, e.g. OSError and ignoring all other interrupts. The worst that could happen then is not having a file system after boot, but only if deliberately interrupted or because e.g. using MU-Editor.

pdg137 commented:
Disabling the interrupt key in _boot_fat.py didn't work well when I tried it.  Catching KeyboardInterrupts and retrying in a loop does work but seems kind of complicated.

> a) should we compensate for the misbehavior of a single PC application, like MU Editor

I think so (though Mu should also get fixed).  What other options do we have for beginners getting started with these devices?  There's Thonny, but it's far more advanced and has many similar bugs of its own.  Mu works well with these simple changes.

> b) should we care about a possible data loss during boot, if Ctrl-C happens at the wrong time.

Seems like obviously yes. For example, if you are using a Pico for data logging, your process for getting the data might be to plug in the board, press Ctrl+C, then run a command to access the data. But if you press Ctrl+C too early it deletes all of your precious data for no reason? That behavior doesn't make sense.

(In my opinion Micropython should NEVER reformat without explicitly prompting the user like any normal device or operating system. But that would be more complicated and require something special for initializing the filesystem the first time.)

dpgeorge commented:
I recently found that I need this exact change on a project I'm working on.  In that project the initial boot code in `_boot.py` takes a while to run and it's very easy to interrupt that code with ctrl-C and leave the hardware in a bad state.  Even with a normal Pico it's pretty easy to hit ctrl-D and ctrl-C fast many times and eventually wipe the filesystem (without warning or notice).

If ctrl-C was only intended for use by a user then maybe the existing behaviour (of being able to interrupt `_boot.py` / frozen start-up code) would be OK.  But ctrl-C is used a lot by automation scripts (eg `mpremote`) and UI's (eg Mu) to get the board into a known state.  If the board is not responding for a short time then it's not possible to know if it's just a slow start up, or an infinite loop in the main application.  The former should not be interrupted, but the latter should.  The only way to distinguish these two cases would be to wait "long enough", and if there's nothing on the serial after "long enough" then assume it's running the application and ctrl-C should break out of it.  But defining "long enough" is impossible for all the different boards.

So I think the solution in this PR is a good one: make it so that frozen start-up code cannot be interrupted by ctrl-C.  That code then effectively acts like normal C start-up code, which IMO is a good thing because it means more start-up code can be moved to Python without any functional change.

Note: on stm32 this was never an issue because all start-up code is in C.  But now other ports start to put more things in `_boot.py` and so we see this problem crop up.

dpgeorge commented on shared/runtime/pyexec.c:
```diff
@@ -113,7 +114,11 @@ STATIC int parse_compile_execute(const void *source, mp_parse_input_kind_t input
         // execute code
-        mp_hal_set_interrupt_char(CHAR_CTRL_C); // allow ctrl-C to interrupt us
+        if (exec_flags & EXEC_FLAG_NO_INTERRUPT) {
+            mp_hal_set_interrupt_char(-1);
```
No need to call this here, it should be `-1` on entry to this function (and if it's not, that's a bug and needs to be fixed elsewhere).  So this can just be:
```c
if (!(exec_flags & EXEC_FLAG_NO_INTERRUPT)) {
    mp_hal_set_interrupt_char(CHAR_CTRL_C);
}
```

dpgeorge commented on shared/runtime/pyexec.h:
```diff
@@ -47,6 +47,7 @@ int pyexec_friendly_repl(void);
 int pyexec_file(const char *filename);
 int pyexec_file_if_exists(const char *filename);
 int pyexec_frozen_module(const char *name);
+int pyexec_frozen_boot_module(const char *name);
```
I think we should have just one function with a flag:
```c
 int pyexec_frozen_module(const char *name, bool allow_keyboard_interrupt);
```

dpgeorge commented:
> If you think this pull request is an improvement, it might also be a good idea to merge pyexec_frozen_module with pyexec_frozen_boot_module , and look at all the places in all the ports where pyexec_frozen_module is called, and figure out which modules should have Ctrl+C enabled during their execution.

Yes I agree.  I think all ports that execute `_boot.py` should not allow ctrl-C during execution of that frozen code.

DavidEGrayson committed:
shared/runtime/pyexec: Don't allow Ctrl+C to interrupt frozen boot code.

Helps prevent the filesystem from getting formatted by mistake, among other
things.  For example, on a Pico board, entering Ctrl+D and Ctrl+C fast many
times will eventually wipe the filesystem (without warning or notice).

Further rationale: Ctrl+C is used a lot by automation scripts (eg mpremote)
and UI's (eg Mu, Thonny) to get the board into a known state.  If the board
is not responding for a short time then it's not possible to know if it's
just a slow start up (eg in _boot.py), or an infinite loop in the main
application.  The former should not be interrupted, but the latter should.
The only way to distinguish these two cases would be to wait "long enough",
and if there's nothing on the serial after "long enough" then assume it's
running the application and Ctrl+C should break out of it.  But defining
"long enough" is impossible for all the different boards and their possible
behaviour.  The solution in this commit is to make it so that frozen
start-up code cannot be interrupted by Ctrl+C.  That code then effectively
acts like normal C start-up code, which also cannot be interrupted.

Note: on the stm32 port this was never seen as an issue because all
start-up code is in C.  But now other ports start to put more things in
_boot.py and so this problem crops up.

Signed-off-by: David Grayson <davidegrayson@gmail.com>

DavidEGrayson commented:
Thanks for the feedback, @dpgeorge !  I force pushed this pull request again to incorporate those changes, and I did some basic tests in the Mu editor with a 3pi+ 2040 Robot to make sure the "Run" and "REPL" buttons are still working properly.  So maybe it's good now.

dpgeorge commented:
Thanks for updating.

The thing to be aware of here is that now, if someone (accidentally) puts an infinite loop in the start-up code, then it's impossible to recover without reflashing the device.  But that should not really be an issue in practice because the frozen start-up is not user-editable, it's provided as part of the board's firmware.  It's a similar level of problem as it would be adding an infinite loop in C to the start-up code.

robert-hh commented:
Should I prepare similar changes for the SAMD, MIMXRT and NRF ports?

DavidEGrayson commented:
In the end this pull request actually applied to all the ports.  You could check my changes and make sure I'm passing the right value for the new allow_keyboard_interrupt argument in those ports.

robert-hh commented:
Thanks. I see.
One note: The double negation logic in pyexec_frozen_module() is a little bit twisted. It's often better to say 'yes' instead of 'not no'.

dpgeorge commented:
> One note: The double negation logic in pyexec_frozen_module() is a little bit twisted. It's often better to say 'yes' instead of 'not no'.

I agree.  But in this case `EXEC_FLAG_NO_INTERRUPT` would need to become `EXEC_FLAG_ALLOW_INTERRUPT` and then all the existing callers of `parse_compile_execute()` would need to be changed.  That flag and function is anyway internal/private to `pyexec.c` so it doesn't matter so much, and can be changed easily without affecting the public API of that file.
